java/arcs/core/util/Lens.kt

```diff
@@ -0,0 +1,187 @@
+ * data structures used by the [Allocator].
+ *
+ * # Lens
+ * A lens is a kind of optic in functional programming that allows immutable data structures to
+ * be updated compactly and elegantly. A [Lens] is constructed around a data structure and
+ * allows its focus target (e.g. field member of a type) to be fetched, modified, or set.
+ *
+ * A [Lens] can be created easily by the [lens] function around a reference to a property, followed
+ * by a code block to copy-construct a class with the modification.
+ *
+ * ## Example
+ * ```kotlin
+ * data class Person(name: String, age: Int)
+ * val nameLens = lens(Person::name) { t,f -> t.copy(name = f) }
+ *
+ * val newPerson = nameLens.set(Person("John", 26), "Tom") // result: Person("Tom", 26)
+ * val name = nameLens.get(newPerson)                      // result: "Tom"
+ * val bang = nameLens.mod(newPerson) { "$it!" }           // result: "Tom!"
+ * ```
+ *
+ * ## Composition
+ * Two or more lens can be composed for deep object graph mutation.
+ * ```kotlin
+ * data class Manager(name: String, employee: Person)
+ *
+ * val employeeLens = lens(Manager::employee) { t,f -> t.copy(employee = f) }
+ * val employeeName = employeeLens + nameLens
+ * val manager = Manager("Sally", person)
+ * val newManager = employeeName.set(manager, "Rita") // Manager("Sally", Person("Rita", 26))
+ * ```
+ *
+ * ## Traversals
+ * When you have a collection field, you need a lens with multiple focii, which is where
+ * [Traversal] comes in. It allows you to mutate each member of a collection field (List or Map)
+ * constructing a new collection. In the case of [Map], the map values are traversed, the keys
+ * are left intact.
+ *
+ * ## Example: Modify the age of each child
+ * ```kotlin
+ * data class Parent(name: String, children: List<Person>)
+ * val parent = Parent("Mary", listOf(Person("Jack", 4), Person("Jill", 3)))
+ * val childrenLens = lens(Parent::children) { t, f = t.copy(children = f) }
+ * val ageLens = lens(Person::age) { t,f -> t.copy(age = f) }
+ *
+ * val allChildren = childrenLens.traverse()
+ * val newParent = allChildren.mod(parent) { it.copy(name = "Baby ${it.name}") }
+ * // newParent = Parent("Mary", listOf(Person("Baby Jack", 4), Person("Baby Jill", 3)))
+ * ```
+ *
+ * ## Traversal Composition
+ * A [Traversal] can be combined with a [Lens], or a another [Traversal] to reach deep into
+ * nested data structures.
+ *
+ * ```kotlin
+ * val allChildrenAges = allChildren + ageLens
+ * val newParent = allChildrenAges.mod(parent) { it + 1 }
+ * // newParent = Parent("Mary", listOf(Person("Jack", 5), Person("Jill", 4)))
+ * ```
+ *
+ * ## Reference
+ * Based on the ideas in
+ * [http://davids-code.blogspot.com/2014/02/immutable-domain-and-lenses-in-java-8.html]
+ *
+ * @property getter a lambda which given [Target] returns [Focus]
+ * @property setter a lambda which given [Target] and [Focus] returns new [Target] set with [Focus]
+ */
+open class Lens<Target, Focus>(
```

Comment:
As I first started reading through the code, I found myself confused for a bit by the typename `Target`. 

When I see the words "focus" and "target", the first sentence that comes to mind is "Focus on the target."

 As I visualized a real lens when thinking about this, it was hard not to thing about the focal point as the, since you are "targeting" a field in the structure. 

I was struggling to come up with a good suggestion to replace it. The best I have so far is `Subject`. When I think about lenses in the context of photography, that language maps really nicely -- e.g. "The subject is the building, the focus is the front door."

Reply:
Subject sounds nice, could also call it Object. Haskell derivations of Lens typically use 't' for everything, where 't' is typeclass, but I didn't want a sea of single-letter symbols.

Reply:
Object definitely makes sense  to me. Although I wonder if some at a glance may first think of `java.lang.Object`.